src/app/modules/auth/auth.controller.ts

```diff
@@ -0,0 +1,195 @@
+import { createLoggerWithLabel } from '../../../config/logger'
+import MailService from '../../services/mail.service'
+import { getRequestIp } from '../../utils/request'
+import * as UserService from '../user/user.service'
+
+import { InvalidOtpError } from './auth.errors'
+import * as AuthService from './auth.service'
+
+const logger = createLoggerWithLabel(module)
+
+/**
+ * Precondition: AuthMiddlewares.validateDomain must precede this handler.
+ * @returns 200 regardless, assumed to have passed domain validation.
+ */
+export const handleCheckUser: RequestHandler = async (_req, res) => {
+  return res.sendStatus(StatusCodes.OK)
+}
+
+/**
+ * Precondition: AuthMiddlewares.validateDomain must precede this handler.
+ */
+export const handleLoginSendOtp: RequestHandler<
+  {},
+  {},
+  { email: string }
+> = async (req, res) => {
+  // Joi validation ensures existence.
+  const { email } = req.body
+  const requestIp = getRequestIp(req)
+  const logMeta = {
+    action: 'handleSendLoginOtp',
+    email,
+    ip: requestIp,
+  }
+
+  // Create OTP.
+  const [otpErr, otp] = await to(AuthService.createLoginOtp(email))
+
+  if (otpErr) {
+    logger.error({
+      message: 'Error generating OTP',
+      meta: logMeta,
+      error: otpErr,
+    })
+    return res
+      .status(StatusCodes.INTERNAL_SERVER_ERROR)
+      .send(
+        'Failed to send login OTP. Please try again later and if the problem persists, contact us.',
+      )
+  }
+
+  // Send OTP.
+  const [sendErr] = await to(
+    MailService.sendLoginOtp({
+      recipient: email,
+      otp: otp!,
+      ipAddress: requestIp,
+    }),
+  )
+  if (sendErr) {
+    logger.error({
+      message: 'Error mailing OTP',
+      meta: logMeta,
+      error: sendErr,
+    })
+
+    return res
+      .status(StatusCodes.INTERNAL_SERVER_ERROR)
+      .send(
+        'Error sending OTP. Please try again later and if the problem persists, contact us.',
+      )
+  }
+
+  // Successfully sent login otp.
+  logger.info({
+    message: 'Login OTP sent successfully',
+    meta: logMeta,
+  })
+
+  return res.status(StatusCodes.OK).send(`OTP sent to ${email}!`)
+}
+
+/**
+ * Precondition: AuthMiddlewares.validateDomain must precede this handler.
+ */
+export const handleLoginVerifyOtp: RequestHandler<
+  {},
+  {},
+  { email: string; otp: string }
+> = async (req, res) => {
+  // Joi validation ensures existence.
+  const { email, otp } = req.body
+
+  const logMeta = {
+    action: 'handleLoginVerifyOtp',
+    email,
+    ip: getRequestIp(req),
+  }
+
+  const [verifyErr] = await to(AuthService.verifyLoginOtp(otp, email))
+
+  if (verifyErr) {
+    logger.warn({
+      message:
+        verifyErr instanceof InvalidOtpError
+          ? 'Login OTP is invalid'
+          : 'Error occurred when trying to validate login OTP',
+      meta: logMeta,
+      error: verifyErr,
+    })
+
+    if (verifyErr instanceof InvalidOtpError) {
+      return res.status(verifyErr.status).send(verifyErr.message)
+    }
```

Comment:
```suggestion
  if (verifyErr) {
    logger.warn({
      message:
        verifyErr instanceof InvalidOtpError
          ? 'Login OTP is invalid'
          : 'Error occurred when trying to validate login OTP',
      meta: logMeta,
      error: verifyErr,

      return res.status(verifyErr.status).send(verifyErr.message)
    }
```
Would this be safer? Although we have tests, I would like to minimise the chance of a future engineer refactoring wrongly or introducing a vulnerability to the login.

Reply:
I will change the check to `instanceof ApplicationError`, and all known errors should be an ApplicationError. However, we will still need the check, since normal errors don't have `err.status`.